### PR TITLE
defer/async load of reCaptcha external script

### PR DESCRIPTION
**Pull request by solracsf**
### Summary of Changes

Load external script with async/defer (as invisble reCaptcha already does):

https://github.com/joomla/joomla-cms/blob/staging/plugins/captcha/recaptcha_invisible/recaptcha_invisible.php#L84

### CODE REVIEW

**Comment by C-Lodder** [comment]
defer **and** async?

Are you sure you know what these attributes do?

**Comment by solracsf** [comment]
https://www.w3.org/TR/2011/WD-html5-20110525/scripting-1.html#attr-script-async

> The defer attribute may be specified even if the async attribute is specified, to cause legacy Web browsers that only support defer (and not async) to fall back to the defer behavior instead of the synchronous blocking behavior that is the default.

https://developers.google.com/recaptcha/docs/display

image

**Comment by C-Lodder** [comment]
@acs-ferreira Last I heard, `async` has a higher priority than `defer` so it would always be used instead. Then `defer` would be used if the browser didn't support `async`

**Comment by solracsf** [comment]
@C-Lodder agree. And it seems that W3C also agrees on that! So we both agree 🎉 💃

**Comment by HLeithner** [comment]
Can you please rebase this PR on j4 branch.

**Comment by twister65** [comment]
Please, can you also modify recaptcha invisible (as you have done in this PR):
https://github.com/joomla/joomla-cms/blob/a81ada410a5bf6b700a79d432fc5926146ac9f94/plugins/captcha/recaptcha_invisible/recaptcha_invisible.php#L72
by
`array('async' => true, 'defer' => true)`

and

https://github.com/joomla/joomla-cms/blob/a81ada410a5bf6b700a79d432fc5926146ac9f94/plugins/captcha/recaptcha_invisible/recaptcha_invisible.php#L84
by
`array('async' => true, 'defer' => true)`

**Comment by twister65** [comment]
When you want to validate an ajax form with a 3rd-party extension, recaptcha is not loaded at all.
Defers activation of the recaptcha widget to the behavior should resolve that. See:
https://www.drupal.org/project/recaptcha/issues/2493183

File: plugins/captcha/recaptcha/recaptcha.php [hunk 1]
```diff
@@ -72,17 +72,35 @@ public function onInit($id = 'dynamic_recaptcha_1')
 			$theme  = $this->params->get('theme', 'clean');
 			$file   = 'https://www.google.com/recaptcha/api/js/recaptcha_ajax.js';
 
-			JHtml::_('script', $file);
+			JHtml::_(
+				'script',
+				$file,
+				array(),
+				array(
+					'defer' => true,
```
Comment (twister65) [inline plugins/captcha/recaptcha/recaptcha.php:80]: `'defer' => 'defer',`

File: plugins/captcha/recaptcha/recaptcha.php [hunk 2]
```diff
@@ -72,17 +72,35 @@ public function onInit($id = 'dynamic_recaptcha_1')
...
+					'async' => true
```
Comment (twister65) [inline plugins/captcha/recaptcha/recaptcha.php:81]: `'async' => 'async'`

File: plugins/captcha/recaptcha/recaptcha.php [hunk 3]
```diff
@@ -72,17 +72,35 @@ public function onInit($id = 'dynamic_recaptcha_1')
...
 			// Load callback first for browser compatibility
-			JHtml::_('script', 'plg_captcha_recaptcha/recaptcha.min.js', array('version' => 'auto', 'relative' => true));
+			JHtml::_(
+				'script',
+				'plg_captcha_recaptcha/recaptcha.min.js',
+				array(
+					'version' => 'auto',
+					'relative' => true
+				), array(
+					'defer' => true,
```
Comment (twister65) [inline plugins/captcha/recaptcha/recaptcha.php:97]: `'defer' => 'defer',`

File: plugins/captcha/recaptcha/recaptcha.php [hunk 4]
```diff
@@ -72,17 +72,35 @@ public function onInit($id = 'dynamic_recaptcha_1')
...
+					'async' => true
```
Comment (twister65) [inline plugins/captcha/recaptcha/recaptcha.php:98]: `'async' => 'async'`

**Review by twister65 — CHANGES_REQUESTED** [review]
> Load external script with async/defer (as invisble reCaptcha already does).

**Comment by hardik-codes** [comment]
> Can you please rebase this PR on j4 branch.

@HLeithner may I do a separate PR for J4?

**Comment by HLeithner** [comment]
> > Can you please rebase this PR on j4 branch.
> 
> @HLeithner may I do a separate PR for J4?

yes a new pr would work better for you

**Comment by brianteeman** [comment]
Pretty sure that the reason for asking if it could be a separate PR was so that this one could also be merged in j3

**Comment by HLeithner** [comment]
oh, if this is the case no, because commits to j3 get merged to j4 by george at the moment.

**Comment by brianteeman** [comment]
which is why I didnt understand the request to rebase

**Comment by hardik-codes** [comment]
@brianteeman @HLeithner is a separate PR needed?

**Comment by HLeithner** [comment]
I missed that this is a bug fix and not a "feature". @hardik-codes forget the rebase or another pr we only need 2 successful tests and I can merge it.

**Comment by hardik-codes** [comment]
Ok
Thanks for the info

**Comment by ghost** [comment]
@acs-ferreira can you please give Test Instructions in first Comment? Maybe we get then Tests.

File: plugins/captcha/recaptcha/recaptcha.php [hunk 5]
```diff
@@ -72,17 +72,35 @@ public function onInit($id = 'dynamic_recaptcha_1')
...
+				)
+			);
 
 			$file = 'https://www.google.com/recaptcha/api.js?onload=JoomlaInitReCaptcha2&render=explicit&hl=' . JFactory::getLanguage()->getTag();
-			JHtml::_('script', $file);
+			JHtml::_('script', $file, array(), array('defer' => true, 'async' => true));
```
Comment (twister65) [inline plugins/captcha/recaptcha/recaptcha.php:103]: `JHtml::_('script',` $file, array('async' => 'async', 'defer' => 'defer'));
As:
https://github.com/joomla/joomla-cms/blob/e7582660d8c160bd842a96841757924e9d0016fa/plugins/captcha/recaptcha_invisible/recaptcha_invisible.php#L84

**Review by twister65 — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
> @acs-ferreira can you please give Test Instructions in first Comment? Maybe we get then Tests.

closed due the lack of response.

**Pull request closed without merging**